design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,398 @@
+```php
+return [
+    'services' => [
+        'Magento\Catalog' => ['address' => 'http://catalog.mydomain.com/'],
+        'Magento\Checkout' => ['address' => 'https://checkout.mydomain.com/'],
+        'Magento\Customer' => ['address' => 'http://192.168.1.1/'],
+    ],
+];
+```
+The invoker will look up the information based on provided services' interface names.
+ 
+#### Changes to RESTful Web API mechanism
+In services communications we would have to send and receive additional metadata (like forwarding ID of the user who
+originally made the request to BFF etc, providing more detailed information about occurred exceptions) so we'd have to
+clearly distinguish requests from other services from requests from front-ends and 1st thing to do is to accept
+inter-service requests on a different URL - _/rest/inter/*_. A controller will be created similar to
+Magento\Webapi\Controller\Rest to process these requests, do specific to service communications validations and to
+initiate unique app area.
+
+Since we declare service contracts with interfaces in \*Api modules and in a monolith deployment every module can use
+other module's service contracts when /Api/* interfaces are available, remote service contracts will be always available
+for other service nodes as well as long as client nodes have _\*Api_ and _\*Proxy_ installed. Remote service contracts
+will be hidden from public access by being either deployed in a private network or by signing requests to each other.
+
+When calling the invoker only interface names and methods names will be given (because those are the service declarations)
+and it would be hard to determine a manually created URLs for endpoints assigned to requested services. Also, except for
+services written using other technologies than Magento, all the requests to services will generated automatically by
+invoker. Therefore there is no need to assign URLs to service contracts manually - they can be generated automatically
+based on interfaces/methods names.
+
+HTTP methods in RESTful API are used to provide self-documentation for developers writing clients and for browsers to
+adjust their behaviour when processing requests to RESTful APIs. They also can be used for caching rules for web
+servers/proxies like Varnish. In services communications requests are being generated automatically, they don't
+come through browsers and caching rules for servers must be assessed on a case-to-case basis and cache needs to be
+carefully invalidated to avoid breaking services' business logic, therefore there is no need in HTTP methods for
+services communications and invoker will always send POST requests to remote services. Services will accept any
+HTTP method though so developers of services not based on Magento can send requests with methods they feel
+suite a requests best.
+
+While URLs for services will be autogenerated and POST always used to make it easier for Magento's invoker it will still
```

Comment:
Seems like POST is used to make invoker implementation easier but communication shouldn't depend on implementation details.